R/generic.R

```diff
@@ -134,3 +136,33 @@ print.R7_generic <- function(x, ...) {
 
   cat(sprintf("<R7_generic> %s with %i methods:\n%s", formals, length(ms), msg), sep = "")
 }
+
+check_generic <- function(fun) {
+  if (!is.function(fun)) {
+    stop("`fun` must be a function", call. = FALSE)
+  }
+
+  if (!has_call(body(fun), quote(method_call))) {
+    stop("`fun` must contain a call to `method_call()`", call. = FALSE)
+  }
+}
+has_call <- function(x, name) {
+  if (!is.call(x)) {
+    return(FALSE)
+  }
+
+  if (identical(x[[1]], name)) {
+    return(TRUE)
+  }
+
+  if (length(x) == 1) {
+    return(FALSE)
+  }
+
+  for (i in seq(2, length(x))) {
```

Comment:
It is not possible for `body(fun)` to return length 0 right? (and then mess up this for loop)

Reply:
I don't think so; a call has to be at least length 1 (the function being called)

Reply:
One case I thought of is for primitive functions, e.g.

``` r
body(sum)
#> NULL
```

<sup>Created on 2021-11-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Not sure if that case would ever come up for this usage, but it might be worth adding a conditional for 0 length just in case?

Reply:
Oh yeah, good catch. I'll just change the `length(x) == 1` check above

Reply:
Although in that case the check for `is.call()` above will have failed.